From: Aron Xu <aron@debian.org>
Date: Mon, 13 Feb 2012 15:16:04 +0800
Subject: [PATCH] quit timer

---
 netcat.c |   29 +++++++++++++++++++++++++++--
 1 files changed, 27 insertions(+), 2 deletions(-)

--- a/netcat.c
+++ b/netcat.c
@@ -75,6 +75,7 @@
 #include <errno.h>
 #include <netdb.h>
 #include <poll.h>
+#include <signal.h>
 #include <stdarg.h>
 #include <stdio.h>
 #include <stdlib.h>
@@ -109,6 +110,7 @@
 int	nflag;					/* Don't do name look up */
 char   *Pflag;					/* Proxy username */
 char   *pflag;					/* Localport flag */
+int     qflag = -1;                             /* Quit after some secs */
 int	rflag;					/* Random ports flag */
 char   *sflag;					/* Source Address */
 int	tflag;					/* Telnet Emulation */
@@ -147,6 +149,7 @@
 
 static int connect_with_timeout(int fd, const struct sockaddr *sa,
         socklen_t salen, int ctimeout);
+static void quit();
 
 int
 main(int argc, char *argv[])
@@ -170,7 +173,7 @@
 	sv = NULL;
 
 	while ((ch = getopt(argc, argv,
-	    "46DdhI:i:jklnO:P:p:rSs:tT:UuV:vw:X:x:z")) != -1) {
+	    "46DdhI:i:jklnO:P:p:q:rSs:tT:UuV:vw:X:x:z")) != -1) {
 		switch (ch) {
 		case '4':
 			family = AF_INET;
@@ -222,6 +225,11 @@
 		case 'p':
 			pflag = optarg;
 			break;
+                case 'q':
+			qflag = strtonum(optarg, INT_MIN, INT_MAX, &errstr);
+			if (errstr)
+				errx(1, "quit timer %s: %s", errstr, optarg);
+			break;
 		case 'r':
 			rflag = 1;
 			break;
@@ -904,7 +912,13 @@
 			}
 			else if (pfd[1].revents & POLLHUP) {
 			shutdown_wr:
-				shutdown(nfd, SHUT_WR);
+			/* if user asked to die after a while, arrange for it */
+			if (qflag > 0) {
+				 signal(SIGALRM, quit);
+				 alarm(qflag);
+			} else {
+				 shutdown(nfd, SHUT_WR);
+                        }
 				pfd[1].fd = -1;
 				pfd[1].events = 0;
 			}
@@ -1138,6 +1152,7 @@
 	\t-O length	TCP send buffer length\n\
 	\t-P proxyuser\tUsername for proxy authentication\n\
 	\t-p port\t	Specify local port for remote connects\n\
+        \t-q secs\t quit after EOF on stdin and delay of secs\n\
 	\t-r		Randomize remote ports\n\
 	\t-S		Enable the TCP MD5 signature option\n\
 	\t-s addr\t	Local source address\n\
@@ -1166,3 +1181,13 @@
 	if (ret)
 		exit(1);
 }
+
+/*
+ * quit()
+ * handler for a "-q" timeout (exit 0 instead of 1)
+ */
+static void quit()
+{
+        /* XXX: should explicitly close fds here */
+        exit(0);
+}
