From: Aron Xu <aron@debian.org>
Date: Mon, 13 Feb 2012 16:09:45 +0800
Subject: [PATCH] help version exit

---
 netcat.c |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

--- a/netcat.c
+++ b/netcat.c
@@ -1234,6 +1234,9 @@
 void
 help(void)
 {
+# if defined(DEBIAN_VERSION)
+        fprintf(stderr, "OpenBSD netcat (Debian patchlevel " DEBIAN_VERSION ")\n");
+# endif
 	usage(0);
 	fprintf(stderr, "\tCommand Summary:\n\
 	\t-4		Use IPv4\n\
@@ -1265,7 +1268,7 @@
 	\t-x addr[:port]\tSpecify proxy address and port\n\
 	\t-z		Zero-I/O mode [used for scanning]\n\
 	Port numbers can be individual or ranges: lo-hi [inclusive]\n");
-	exit(1);
+	exit(0);
 }
 
 void
