From: Aron Xu <aron@debian.org>
Date: Tue, 14 Feb 2012 23:02:00 +0800
Subject: [PATCH] serialized handling multiple clients

---
 netcat.c |   13 ++++++-------
 1 files changed, 6 insertions(+), 7 deletions(-)

--- a/netcat.c
+++ b/netcat.c
@@ -432,14 +432,13 @@
 				s = unix_bind(host);
 			else
 				s = unix_listen(host);
-		}
+		} else
+			s = local_listen(host, uport, hints);
+		if (s < 0)
+			err(1, NULL);
 
 		/* Allow only one connection at a time, but stay alive. */
 		for (;;) {
-			if (family != AF_UNIX)
-				s = local_listen(host, uport, hints);
-			if (s < 0)
-				err(1, NULL);
 			/*
 			 * For UDP, we will use recvfrom() initially
 			 * to wait for a caller, then use the regular
@@ -506,15 +505,16 @@
 				close(connfd);
 			}
 
-			if (family != AF_UNIX)
+			if (kflag)
+				continue;
+			if (family != AF_UNIX) {
 				close(s);
+			}
 			else if (uflag) {
 				if (connect(s, NULL, 0) < 0)
 					err(1, "connect");
 			}
-
-			if (!kflag)
-				break;
+			break;
 		}
 	} else if (family == AF_UNIX) {
 		ret = 0;
