--- a/netcat.c
+++ b/netcat.c
@@ -115,6 +115,8 @@
 
 #define UDP_SCAN_TIMEOUT 3			/* Seconds */
 
+#define LISTEN_BACKLOG 5
+
 /* Command Line Options */
 int	bflag;					/* Allow Broadcast */
 int     Cflag = 0;                              /* CRLF line-ending */
@@ -818,7 +820,7 @@
 	if ((s = unix_bind(path)) < 0)
 		return (-1);
 
-	if (listen(s, 5) < 0) {
+	if (listen(s, LISTEN_BACKLOG) < 0) {
 		close(s);
 		return (-1);
 	}
@@ -1090,7 +1092,7 @@
 	} while ((res0 = res0->ai_next) != NULL);
 
 	if (!uflag && s != -1) {
-		if (listen(s, 1) < 0)
+		if (listen(s, LISTEN_BACKLOG) < 0)
 			err(1, "listen");
 	}
 
